### PR TITLE
Prometheus schematization

### LINKED ISSUE #63874: Prometheus schematization
The guidance https://docs.google.com/document/d/1RRPGeD7DQ7wot0Pwnc4UZolXrZYip0ViqOozl9zbBpE/edit#

### PR DESCRIPTION
**Pull request by itsmylife**
**What is this feature?**

This is an example PR to illustrate schematizing the Prometheus datasource plugin. Probably It won't be merged.
The guidance https://docs.google.com/document/d/1RRPGeD7DQ7wot0Pwnc4UZolXrZYip0ViqOozl9zbBpE/edit#

**Which issue(s) does this PR fix?**:

Closes https://github.com/grafana/grafana/issues/63874

**Special notes for your reviewer**:
This PR is an up-to-date version of [this PR](https://github.com/grafana/grafana/pull/61872). I closed that one because it was behind the main and getting hard to merge it because of the conflicts.

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/dataquery.cue [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+					{
+						common.DataQuery
+						expr:            string
+						format?:         string
+						instant?:        bool
+						range?:          bool
+						exemplar?:       bool
+						hinting?:        bool
+						interval?:       string
+						intervalMs?:     int64
```
Comment (ryantxu) [inline public/app/plugins/datasource/prometheus/dataquery.cue:41]: Are these set in the UI, or as part of the request?  If we can not set them from the UI, lets keep them out of this schema.

File: public/app/plugins/datasource/prometheus/dataquery.cue [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	maturity: "merged"
+
+	lineage: {
+		seqs: [
+			{
+				schemas: [
+					{
+						common.DataQuery
+						expr:            string
+						format?:         string
```
Comment (ryantxu) [inline public/app/plugins/datasource/prometheus/dataquery.cue:35]: What are the valid values?  I think 'timeseries' and 'table', but would be nice to have that explicit :)

File: public/app/plugins/datasource/prometheus/dataquery.cue (hunk 1 above)
Comment (sdboyer) [reply public/app/plugins/datasource/prometheus/dataquery.cue:41]: while i have a feeling we may find exceptions to this rule (such that we need to find a better rule), i think it's the right one at least for now

File: pkg/tsdb/prometheus/models/query.go [hunk 3]
```diff
@@ -45,16 +45,12 @@ const (
...
-	InstantQuery   bool   `json:"instant"`
-	ExemplarQuery  bool   `json:"exemplar"`
-	IntervalFactor int64  `json:"intervalFactor"`
-	UtcOffsetSec   int64  `json:"utcOffsetSec"`
+	dataquery.PrometheusDataQuery
+	LegendFormat *string `json:"legendFormat,omitempty"`
+	RequestId    *string `json:"requestId,omitempty"`
+	// Timezone offset to align start & end time on backend
+	UtcOffsetSec   *int64 `json:"utcOffsetSec,omitempty"`
+	ValueWithRefId *bool  `json:"valueWithRefId,omitempty"`
```
Comment (sdboyer) [inline pkg/tsdb/prometheus/models/query.go:53]:
```suggestion
	LegendFormat string `json:"legendFormat,omitempty"`
	RequestId    string `json:"requestId,omitempty"`
	// Timezone offset to align start & end time on backend
	UtcOffsetSec   int64 `json:"utcOffsetSec,omitempty"`
	ValueWithRefId bool  `json:"valueWithRefId,omitempty"`
```

It's probably fine to leave these as non-pointers, same as they were before. Codegen produces pointers to represent field optionality (see grafana/grok#1) in the schema. We may not even preserve that behavior long term because of how annoying it is to work with pointers to certain primitive types in Go.

File: public/app/plugins/datasource/prometheus/dataquery.cue (hunk 2 above)
Comment (sdboyer) [reply public/app/plugins/datasource/prometheus/dataquery.cue:35]: Agree, if there's a set of acceptable values here, they should be explicitly declared. 

Could make it a reference to another type, as with `#QueryEditorMode`, if that type needs to be used elsewhere in TS

**Review by sdboyer — CHANGES_REQUESTED** [review]
i don't know the specifics of the prom datasource, but overall this looks pretty close! i think i'd just simplify a bit by removing pointers and some possibly-unnecessary fields from the schema as @ryantxu noted

File: public/app/plugins/datasource/prometheus/dataquery.cue (hunk 2 above)
Comment (itsmylife) [reply public/app/plugins/datasource/prometheus/dataquery.cue:35]: I will check the possible values and introduce a change.

Comment (itsmylife) [reply public/app/plugins/datasource/prometheus/dataquery.cue]: See https://github.com/grafana/grafana/pull/63878/commits/a0bd6216c63b03672a39ed102f498cc213514b49

File: public/app/plugins/datasource/prometheus/dataquery.cue (hunk 1 above)
Comment (itsmylife) [reply public/app/plugins/datasource/prometheus/dataquery.cue]: I made the changes related to this.

File: docs/sources/developers/kinds/composable/prometheusdataquery/schema-reference.md [hunk 4]
```diff
@@ -15,25 +15,19 @@ title: PrometheusDataQuery kind
...
+| `hide`       | boolean | No       | *(Inherited from [DataQuery](#dataquery))*<br/>true if query is disabled (ie should not be returned to the dashboard)                                                                                                                                                                                  |
+| `instant`    | boolean | No       |                                                                                                                                                                                                                                                                                                        |
+| `key`        | string  | No       | *(Inherited from [DataQuery](#dataquery))*<br/>Unique, guid like, string used in explore mode                                                                                                                                                                                                          |
```
Comment (ryantxu) [inline docs/sources/developers/kinds/composable/prometheusdataquery/schema-reference.md:28]: I was surprised to see "key" here... but then see that is not added in this PR

https://github.com/grafana/grafana/pull/64467

File: public/app/plugins/datasource/prometheus/dataquery.cue [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+composableKinds: DataQuery: {
+	maturity: "merged"
+
+	lineage: {
+		seqs: [
+			{
+				schemas: [
+					{
+						common.DataQuery
```
Comment (ryantxu) [inline public/app/plugins/datasource/prometheus/dataquery.cue:33]: Any effort on comments here will be HUGELY helpful.  The text+values here will drive all the docs and API generation, so if we can try to explain the parameters here (and agree/confirm they are accurate!) that would be awesome.

File: public/app/plugins/datasource/prometheus/dataquery.cue [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+						expr:        string
+						instant?:    bool
+						range?:      bool
+						exemplar?:   bool
```
Comment (ryantxu) [inline public/app/plugins/datasource/prometheus/dataquery.cue:37]:
```suggestion
                                                  // Execute an additional query to identify interesting raw samples relevant for the given expr
						exemplar?:   bool
```

File: public/app/plugins/datasource/prometheus/dataquery.cue [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+	lineage: {
+		seqs: [
+			{
+				schemas: [
+					{
+						common.DataQuery
+						expr:        string
+						instant?:    bool
+						range?:      bool
```
Comment (ryantxu) [inline public/app/plugins/datasource/prometheus/dataquery.cue:36]: Can both range and instant be true?

File: pkg/tsdb/prometheus/models/query_test.go [hunk 8]
```diff
@@ -27,6 +27,8 @@ func TestParse(t *testing.T) {
 		queryJson := `{
 			"expr": "go_goroutines",
 			"refId": "A",
+			"interval": "1s",
+			"intervalMs": 1000,
```
Comment (ryantxu) [inline pkg/tsdb/prometheus/models/query_test.go:31]: Are these changes still necessary?

File: pkg/tsdb/prometheus/models/query.go [hunk 9]
```diff
@@ -45,16 +45,13 @@ const (
...
-	LegendFormat   string `json:"legendFormat"`
-	Interval       string `json:"interval"`
-	IntervalMS     int64  `json:"intervalMS"`
-	StepMode       string `json:"stepMode"`
-	RangeQuery     bool   `json:"range"`
-	InstantQuery   bool   `json:"instant"`
-	ExemplarQuery  bool   `json:"exemplar"`
-	IntervalFactor int64  `json:"intervalFactor"`
-	UtcOffsetSec   int64  `json:"utcOffsetSec"`
+	dataquery.PrometheusDataQuery
```
Comment (ryantxu) [inline pkg/tsdb/prometheus/models/query.go:48]:
```suggestion
	dataquery.PrometheusDataQuery
	
	// The following properties may be part of the request payload, however they are not saved in panel JSON
```

File: public/app/plugins/datasource/prometheus/dataquery.cue (hunk 7 above)
Comment (sdboyer) [reply public/app/plugins/datasource/prometheus/dataquery.cue:36]: they can't, and we can natively constrain them in CUE to indicate that invariant, but at the moment cuetsy evaluates the operation and converts the TS output to just `false` - see grafana/cuetsy#75

Comment (itsmylife) [reply public/app/plugins/datasource/prometheus/dataquery.cue:36]: Yes, it can be. In query options there is a `type` option field:
image

File: pkg/tsdb/prometheus/models/query_test.go (hunk 8 above)
Comment (itsmylife) [reply pkg/tsdb/prometheus/models/query_test.go:31]: Let me check them once more. Probably we won't need them anymore. Good catch.

**Review by ryantxu — APPROVED** [review]
@itsmylife -- this looks like a great step forward.  My only remaining comments would be to improve the comments so that a noice looking at the fields+comments could understand what they do.

But that can also evolve as we have more people look at and use this :)

File: public/app/plugins/datasource/prometheus/dataquery.cue (hunk 7 above)
Comment (sdboyer) [reply public/app/plugins/datasource/prometheus/dataquery.cue:36]: oh! i totally misunderstood this on my first pass through it, then :)

File: public/app/plugins/datasource/prometheus/dataquery.cue [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	common "github.com/grafana/grafana/packages/grafana-schema/src/common"
+	"github.com/grafana/grafana/pkg/plugins/pfs"
+)
+
+// This file (with its sibling .cue files) implements pfs.GrafanaPlugin
+pfs.GrafanaPlugin
+
+composableKinds: DataQuery: {
+	maturity: "merged"
```
Comment (sdboyer) [inline public/app/plugins/datasource/prometheus/dataquery.cue:26]:
```suggestion
	maturity: "experimental"
```

**Review by sdboyer — APPROVED** [review]
agreed with @ryantxu, i think this is good enough to go in - and be marked as `experimental`. While it's absolutely true that the comments here are crucial for the as-code user use case, it takes a different mindset and knowledge base than just getting the types hooked up. It's OK to plan for a follow-up to improve them. 

Thanks for all your work on this!

**Review by gtk-grafana — APPROVED** [review]
LGTM, thanks for all the hard work on this, I'm excited to have type synchronization across the stacks!

**Review by bohandley — APPROVED** [review]
LGTM! The comments make sense to me as well, good work!
